Feature: Add support for Spanish

canasmh opened the pull request:
Relevant issue: #4

# Description

### Brief Overview
This PR adds a `TARGET_LANGUAGE` key to the `.env` file. By default the value of `TARGET_LANGUAGE` is Japanese, but Spanish is also a valid option. It also replaces all instances of `'Japanese'` with `settings.TARGET_LANGUAGE`. As a result, the tests file for the `word_service` function had to be modified to ensure the correct language is being stored. Furthermore, the pormpt for the OpenAI was slightly modified to only return furigana _if_ it applies to that language

### Tests
The `test_word_service` file was modified so that the `StoreWordTests` class uses the environment variable to confirm that the correct language was stored in the database. All tests are go ✅

### Example
Below is a screenshot of the email I received from ichikoto, with `TARGET_LANGUAGE="Spanish"`:
<img width="495" alt="Screenshot 2023-11-20 at 6 44 18 PM" src="https://github.com/MattTuccillo/ichikoto/assets/55603689/d0c5942c-0aea-45ec-83c6-3c68b122b174">


### Disclaimers
I was a little late in seeing that no changes should be made to `setup.py`. My approach to adding support for Spanish while still keeping Japanese the default was by modifying `setup.py` and adding `TARGET_LANGUAGE="Japanese"` to the `.env` file. I think this approach would make it easier to add support for even more languages, while keeping the source code relatively clean of too many conditionals for the language used. 

It also reduces the complexity of the `store_word` function by removing the need for a second argument. Nonetheless, if the PR is disapproved for this reason I will be happy to visit other avenues where `setup.py` is not modified

canasmh committed:
Initial commit

Add target language to ENV file and extract it in settings.py

canasmh committed:
Use environment variable for language

With the environment variable available, we no longer need to pass in a language param to store_word

canasmh committed:
Use env variable for target language in prompt

This requires slight modification to the prompt, since by default we are asking for furigana, which won't be applicable for Spanish

canasmh committed:
Fix typo

'Japanese' instead of 'Japenese'

canasmh committed:
Modify current word_service test

The test now pulls the target language from env and verifies the correct language is stored

MattTuccillo commented:
I will clone your branch tomorrow and take a look at things a little closer but it's looking good so far. No worries about the setup.py, I mainly added that incase someone decided to add French first, I didn't want two tickets with the same criteria. I just more-so wanted to see the support system come together and get confirmation that Spanish will work.  Great work though, thanks!

MattTuccillo commented on ichikoto/settings.py:
```diff
@@ -158,6 +158,15 @@
 else:
     EMAIL_SCHEDULER_MINUTES = 0
 
+# Grabs the users target language
+language = os.getenv("TARGET_LANGUAGE")
+available_languages = ["Japanese", "Spanish"]
+
+if language.capitalize().strip() in available_languages:
+    TARGET_LANGUAGE = language.capitalize().strip()
+else:
+    TARGET_LANGUAGE = available_languages[0]
+
```
well planned, meant to add an issue to standardize the language names, nice!

canasmh replied:
Yeah I figured you'd only want to allow languages that have already been through some sort of PR/QA stage, so..

MattTuccillo commented on word_manager/services/gpt_service.py:
```diff
@@ -25,17 +25,17 @@ def send_prompt(manual_mode=False, simulate_error=None):
-        content_message += f"\nResponse Structure:\nWord: [In Japanese]\n"
-        content_message += f"Definition: [In Japanese with furigana]\n"
-        content_message += f"Context Sentence: [In Japanese with furigana]\n"
+        content_message += f".\nDefinition in {settings.TARGET_LANGUAGE}: Give the definition of the word in {settings.TARGET_LANGUAGE} with furigana (if applicable).\n"
+        content_message += f"Context Sentence in {settings.TARGET_LANGUAGE}: Offer a sentence using the word in {settings.TARGET_LANGUAGE} with furigana (if applicable).\n"
+        content_message += f"\nResponse Structure:\nWord: [In {settings.TARGET_LANGUAGE}]\n"
+        content_message += f"Definition: [In {settings.TARGET_LANGUAGE} with furigana (if applicable)]\n"
+        content_message += f"Context Sentence: [In {settings.TARGET_LANGUAGE} with furigana (if applicable)]\n"
```
Try and rework this a tad. I'd like something like a conditional for the furigana part, maybe even a switch case based on the languages if you want to support the possibility of more language nuances.

So to summarize, just add a conditional that omits the 'with furigana' parts if it's not Japanese and then apply the period as usual.

MattTuccillo replied:
This is probably outside of AC, but you went the extra mile to fully implement Spanish so lets take it all the way home haha

canasmh replied:
Awesome, I just pushed commit [2b248a0](https://github.com/MattTuccillo/ichikoto/pull/10/commits/2b248a0c4f9899483057faddf37ea848e1187ff3) that I think addresses this issue.

MattTuccillo reviewed (changes requested):
Ok this comment was large before cause I wasn't sure how it would all look after finishing the review. Everything looks good though, only a minor change below dealing with the prompt sending. Otherwise looking good to me! GJ!

canasmh committed:
Create function to handle language-specific instructions

For now, the function returns ' with furigana' if lang=Japanese, else returns ''

MattTuccillo reviewed (approved):
LGTM! Great work and thanks for making the fix!
